### PR TITLE
Fix for the Never-Consent dialog showing up repeatedly on the same site

### LINKED ISSUE #908: Never-Consent popup is shown multiple times for the same page
If you open a website that is supported by [Never Consent](https://www.ghostery.com/blog/never-consent-by-ghostery-new-feature-removing-annoying-cookie-pop-ups-automates-interaction-with-consent-dialogs), it will initially show a dialog like this:
never-consent

Once you selected "Yes" or "No", it should remember and not ask about the same page again. If it is shown multiple times for the same website, it is a bug. For instance, it can be reproduced on forbes.com: https://github.com/ghostery/broken-page-reports/issues/119

Breakage on individual pages are best reported on https://github.com/ghostery/broken-page-reports, but this issue here seems to be generic; not clear yet what happens, but the user choice does not seem to be reliably remembered by the extension.

---

Background: Last week, Never-Consent was released on Opera/Yandex (2022-10-17) and Edge (2022-10-19). We are working on releasing it soon on Chrome and Firefox, but today (2022-10-24) is not live yet.

### PR DESCRIPTION
**Pull request by philipp-classen**
The URL normalization differed between content_script and background:
* https://github.com/ghostery/ghostery-extension/blob/94cfdde011b3f4f8310dac429bc740feb49ea7fb/extension-manifest-v2/app/content-scripts/autoconsent.js#L27
* https://github.com/ghostery/ghostery-extension/blob/94cfdde011b3f4f8310dac429bc740feb49ea7fb/extension-manifest-v2/src/modules/autoconsent.js#L29

Thus, "https://www.forbes.com/" was stored as "forbes.com" but later checked against "www.forbes.com".

Fixes #908

### CODE REVIEW

File: extension-manifest-v2/src/modules/autoconsent.js [hunk 1]
```diff
@@ -15,41 +15,42 @@ import globals from '../classes/Globals';
...
+	}
 
-	if (
-		enable_autoconsent &&
-		!globals.SESSION.paused_blocking &&
-		!autoconsent_blacklist?.includes(url.hostname) &&
-		!site_whitelist.includes(url.hostname.replace(/^www\./, ''))
-	) {
-		const optOut = autoconsent_whitelist?.some(h => url.hostname.includes(h)) ?? true;
+	const url = new URL(msg.url);
```
Comment (chrmod) [inline extension-manifest-v2/src/modules/autoconsent.js:28]: maybe use tldts-experimental or cliqz/url-parser instead?

**Review by chrmod — APPROVED** [review]

**Review by smalluban — COMMENTED** [review]
Looks to me, that the condition logic is the same, but put differently. What is the clue of the PR?

File: extension-manifest-v2/src/modules/autoconsent.js [hunk 2]
```diff
@@ -15,41 +15,42 @@ import globals from '../classes/Globals';
...
+	const site = url.hostname.replace(/^www\./, '');
```
[the remark this answers is not in the document]
Comment (smalluban) [reply extension-manifest-v2/src/modules/autoconsent.js:29]: This is the way how the `site_whitelist` works. I found that clean up somewhere in the code. The problem is that our content script must pass the `host` into the popup iframe, so I would stick with that.

[the remark this answers is not in the document]
Comment (philipp-classen) [reply extension-manifest-v2/src/modules/autoconsent.js:29]: First, I was experimenting with using a different normalization function (e.g. google.com and google.de could be put in the same bucket). But then I also realized that the hostname plus leading www removal is used in various places in the code.

Like Dominik said: breaking consistency with the trusted pages, which is stored in site_whitelist, opens the possibility of weird edge (e.g. trusting pages behaving slightly different from Never-Consent site lists). That is why also ended up keeping the existing mechanism unchanged.

File: extension-manifest-v2/src/modules/autoconsent.js (hunk 1 above)
Comment (philipp-classen) [reply extension-manifest-v2/src/modules/autoconsent.js:28]: Again, I would try to stay as close as possible to the other normalization functions, which are using URL in the end (content script and also the whitelist code, which uses parseUrl).

**Review by smalluban — APPROVED** [review]

**Comment by philipp-classen** [comment]
> Looks to me, that the condition logic is the same, but put differently. What is the clue of the PR? Only the `www` problem? If so, we could just clean the url in one place ..

I would argue that it easier to read. But strictly speaking it is not fully equivalent:

```
const optOut = autoconsent_whitelist?.some(h => url.hostname.includes(h)) ?? true;
```

and

```
const globallyEnabled = !autoconsent_whitelist;
const optOut = globallyEnabled || autoconsent_whitelist.includes(site);
````

The first may have some false-positive, e.g. after whitelisting "z.test", it would match "xyz.test".

**Pull request merged**